inst/stan/functions/discretised_logit_hazard.stan

```diff
@@ -42,54 +44,98 @@ vector adjust_lcdf_discrete(vector lcdf, int n, int max_strat) {
+    // If n were infinite this would be equivalent to dividing by 2.
+    if (n == 1) {
+      adjusted_lcdf = lcdf - lcdf[1];
+    }
+    if (n > 1) {
+      adjusted_lcdf = lcdf - log_sum_exp(lcdf[n], lcdf[n-1]);
+    }
   } else {
     reject("Unknown strategy to handle probability mass beyond the maximum value.");
   }
   return(adjusted_lcdf);
 }
 
-// compute discretised log scale pmf
-vector lcdf_to_log_prob(vector lcdf, int n) {
+// compute discretised log scale pmf from 0:n
+// assumes double censoring and a uniform interval approximation
+vector lcdf_to_uniform_double_censored_log_prob(vector lcdf, int n) {
   vector[n] lpmf; 
+  // p_0 = cdf_1 - cdf_0
   lpmf[1] = lcdf[1];
-  // p = cdf_n - cdf_n-1
-  lpmf[2:n] = log_diff_exp(lcdf[2:n], lcdf[1:(n-1)]);
+  if (n > 1) {
+    // p_1 = cdf_2 - cdf_0
+    lpmf[2] = lcdf[2];
+    if (n > 2) {
+      // p_n = cdf_n+1 - cdf_n-1
+      lpmf[3:n] = log_diff_exp(lcdf[3:n], lcdf[1:(n-2)]);
+    }
+  }
   return(lpmf);
 }
 
 // compute discretised logit hazard
-vector lcdf_to_logit_hazard(vector lcdf, int n) {
+vector lprob_to_uniform_double_censored_log_hazard(vector lprob, vector lcdf,
+   int n) {
   vector[n] lhaz;
-  vector[n-1] lccdf;
-  // cccdf = 1 - cdf
-  lccdf = log1m_exp(lcdf[1:(n-1)]);
+  // h_0 = cdf_1
   lhaz[1] = lcdf[1];
-  // h = 1 - CCDF_n / CCDF_n-1
-  lhaz[2:(n-1)] = log1m_exp(lccdf[2:(n-1)] - lccdf[1:(n-2)]);
+  // h_n = p_n / (1 - sum^{n-1}_{d=0} p_d)
+  // h_n = cdf_n+1 - cdf_n-1 / (1 - sum^{n-1}_{d=0} cdf_d+1 - cdf_d-1)
+  // h_n = cdf_n+1 - cdf_n-1 / 1 - sum^n_{d=0} cdf_d + sum^n-2_{d=1} cdf_d
+  // h_n = cdf_n+1 - cdf_n-1 / 1 - cdf_n - cdf_n-1
+  // h_n = (cdf_n+1 - cdf_n-1) / (cccdf_n - cdf_n-1)
+  // log(h_n) = log(cdf_n+1 - cdf_n-1) - log(cccdf_n - cdf_n-1)
+  if (n > 1) {
+    vector[n-2] lccdf;
+    // cccdf_n = 1 - cdf_n
+    lccdf = log1m_exp(lcdf[1:(n-2)]);
+    lhaz[2] = lprob[2] - lccdf[1];
+    if (n > 2) {
+      lhaz[3:(n-1)] = lprob[3:(n-1)] - log_diff_exp(lccdf[2:(n-2)], lcdf[1:(n-3)]);
+    }
+  }
+  return(lhaz);
+}
+
+// Convert from log hazard to logit hazard efficiently (i.e. without converting
+// to the natural scale)
+vector log_hazard_to_logit_hazard(vector lhaz, int n) {
+  vector[n] logit_haz;
   // Logit transformation
-  lhaz[1:(n-1)] = lhaz[1:(n-1)] - log1m_exp(lhaz[1:(n-1)]);
+  logit_haz[1:(n-1)] = lhaz[1:(n-1)] - log1m_exp(lhaz[1:(n-1)]);
   // Set last logit transformed hazard to Inf (i.e h[n] = 1)
-  lhaz[n] = positive_infinity();
-  return(lhaz);
+  logit_haz[n] = positive_infinity();
+  return(logit_haz);
 }
 
 // Calculate discreteised logit hazard or log probability
 // up to a maximum oberved delay with a range of normalisation strategies
-// Calculate the daily hazard of reporting using parametric
-// distributions up to the maximum observed delay
+// Assumes that delays are doubly censored and that the interval width is
+// approoximated by a uniform.
 vector discretised_logit_hazard(real mu, real sigma, int n, int dist, 
```

Comment:
small typos:
- approximated
- double-censored?

interval width is approximately uniformly distributed?

Reply:
wow spelling Sam. Yes exactly. The combination of two uniforms is not a uniform so this is only an approximation.